syncer/syncer.go

```diff
@@ -113,13 +117,22 @@ func NewSyncer(ctx context.Context, conf Configuration, ticker layerTicker, mesh
 		awaitSyncedCh:     make([]chan struct{}, 0),
 		shutdownCtx:       shutdownCtx,
 		cancelFunc:        cancel,
+		tg:                taskgroup.New(taskgroup.WithContext(shutdownCtx)),
 	}
 }
 
 // Close stops the syncing process and the goroutines syncer spawns.
 func (s *Syncer) Close() {
-	// TODO: ensure goroutines are all terminated before shutting down
+	s.syncTimer.Stop()
 	s.cancelFunc()
+	if s.getStarted() {
+		s.logger.Info("waiting for goroutines to finish")
+		if err := s.tg.Wait(); err != nil {
```

Comment:
i think error can be ignored if it is a context.Canceled, since it means that it was stopped properly